programs/conditional_vault/src/instructions/merge_tokens.rs

```diff
@@ -48,23 +78,44 @@ impl<'info, 'c: 'info> InteractWithVault<'info> {
             )?;
         }
 
+        // Validate token accounts and decimals
+        require_eq!(
+            accs.user_underlying_token_account.mint,
+            accs.vault_underlying_token_mint.key(),
+            VaultError::UnderlyingTokenMintMismatch
+        );
+        require_eq!(
+            accs.vault_underlying_token_account.mint,
+            accs.vault_underlying_token_mint.key(),
+            VaultError::UnderlyingTokenMintMismatch
+        );
```

Comment:
These are already checked in `InteractWithVault`

Reply:
> These are already checked in `InteractWithVault`

Yes I performed those validations before leveraging the usefulness of TransferChecked and it seems kind of redundant. But even with that, TransferChecked remains a good option.